### ISSUE TITLE
Stuttering Audio on Chrome/Chromium (HTML5, Threads)

### ISSUE THREAD

**Issue by FernandoValcazara** [godotengine/godot#43210]
**Godot version:**
3.2.4 Beta 1

**OS/device including version:**
Windows 10 - GLES2 - Custom Build with Threads

**Issue description:**
Stuttering Audio with Godot 3.2.4 Beta 1 (HTML5 Build)
Unfortunately with all tests we have done so far any build exported for Web have its audios heavily stuttering.

**Steps to reproduce:**
I've made a simple example below for testing purposes, this example has only one audio.

**Minimal reproduction project:**
Source Code:
[AudioTest.zip](https://github.com/godotengine/godot/files/5465593/AudioTest.zip)

Online access:
https://jogosdev.afinando.com.br/AudioTest/

**Comment by akien-mga** [comment]
Is it only in 3.2.4 Beta 4 or does it happen with 3.2.3 stable too?

**Comment by FernandoValcazara** [comment]
It does happen on 3.2.3 stable too, but only with for users with low performance computers, it is a know open issue.

But with 3.2.4 Beta 1 it is happening even with computer of high performance, we don't have users testing the 3.2.4 Beta 1 yet, this performance test I'm talking about was ran on the computers of our development team. Which are high performance computers.

**Comment by Faless** [comment]
I can confirm the issue on Linux, with Chrome/Chromium, but it works fine with Firefox (which is I guess why I didn't spot it).
I'll update the issue, and see what I can do.

**Comment by groud** [comment]
As a side note, nice issue number. :p

**Comment by probrain-dev** [comment]
Unfortunately we're still having stuttering problems in low performance devices when using browsers such as G. Chrome, Edge and Opera. Firefox seems to be the only one that works flawlessly.

Here is an updated version as example.
https://jogosdev.afinando.com.br/AudioTest/

**Comment by Calinou** [comment]
@probrain-dev Did you export the project with Godot 3.2.4beta4 and with threads enabled?

**Comment by probrain-dev** [comment]
I Tested with Regular, Threads and GDNative, and the Output Latency with the vaues 20, 50, 60.
I used the last beta version 3.2.4 beta5.

**Comment by Calinou** [comment]
@probrain-dev Which exact device models did you encounter issues on?

**Comment by probrain-dev** [comment]
We tested a few different game engines, such as Cocos Creator and Defold, and didn't had audio issues on the low performance devices.

But Godots' had issues such as:

**First tested device**: MOTO G Play, Android 7.1.1, Google Chrome
With Threads enabled, the game doesn't even load.
I used a regular build and then the game loaded, but the audio was stuttering.

**Second tested device**: i3 550 2gb ram - Windows 10 (Outdated Computer), Google Chrome
The graphics and audio were stuttering.

But on Firefox everything worked fine.

**Comment by probrain-dev** [comment]
We tested at the new version 3.2.4.rc1and the problem still happen. 
https://jogosdev.afinando.com.br/AudioTestRegular/
https://jogosdev.afinando.com.br/AudioTestThread/

**Devices**:
MOTO G Play, Android 7.1.1, Google Chrome
I3 550 2gb ram - Windows 10 (Outdated Computer), Google Chrome

**Comment by wandomPewlin** [comment]
Have you tried increasing the `Project Settings -> Application -> Audio -> output_latency.web` to more than 70? This is the lowest I could go without crackling with Chrome on Android.

**Comment by Nosliwnayr** [comment]
On a school project I'm working on in 3.2.4 rc1, enabling threads gives me stuttering audio. While a regular build does not.
For software I'm running chrome beta 89.0.4389.40 on PoP os, and my hardware is a laptop with an i7-10710U using an AMD Vega 64 graphics card in an egpu.

[Here](https://youtu.be/eV9Mx3o6_2o) is a capture of the behavior on my machine.
In Firefox the page just shows a `SharedArrayBuffer is not defined` error and won't load it.

The project that gives me stuttering audio is available [here](https://nayru2210.itch.io/asteroids) and the source folder I uploaded to itch is [here](https://github.com/godotengine/godot/files/5963943/build.zip) if that helps, but it has no debug information in it.

Here is a [build with debug information](https://github.com/godotengine/godot/files/5963917/threads_debug.zip), but neither that or the original project will load for me when I try to serve it locally with http-server.
[This](https://github.com/godotengine/godot/files/5963910/127.0.0.1-1613032564466.log) is the error log I got from Chrome, and Firefox displays the same `SharedArrayBuffer is not defined` error and won't load the page.

Let me know if you want the source project, or if you need more specific information about my machine.

**Comment by Faless** [comment]
@Nosliwnayr can you tell me what number gets displayed when you visit https://jsfiddle.net/g5609dk7/ on that machine?
The simple script there should print the AudioContext mix rate, since I have a feeling chrome does not like custom mix rates.

**Comment by probrain-dev** [comment]
@vandonpeulin, that's perfect, we tested with 100 and 150 of output latency, and with 150 the stuttering was 100% gone. Thank you for the advice.

We will be waiting for the stable version so we can release a few new games using Godot's internal audio engine, so we can be sure the problem with low-performance devices is totally gone.

**Comment by Nosliwnayr** [comment]
@Faless on Linux in either Firefox or Chrome I got 44100 on that page.
I rebooted into Windows and in Chrome beta 89.0.4389.40 I get no audio stuttering and the jsfiddle displayed 48000.
Chrome beta 89.0.4389.48 on my Pixel 4 displays the same `SharedArrayBuffer is not defined` when I try to run the project, and the jsfiddle displays 48000.

**Comment by fabriceci** [comment]
I have the same issue on Chrome (88.0.4324.150) the jsfiddle return 44100. Hope this (performance/stuttered) on HTML5 will be improved, that prevent to spread the engine on the web.

**Comment by Faless** [comment]
@fabriceci please also include OS and machine information (CPU/GPU/memory)

**Comment by fabriceci** [comment]
@Faless I restarted my computer and now it's OK on the threaded version, sorry my bad.

**Comment by rcbotCheeseh** [comment]
A lot of my old HTML5 exports are exhibiting this behaviour now when they didn't used to - perhaps a browser update has caused some instability / incompatibility?

**Comment by davcri** [comment]
I can confirm the issue. Made a minimal reproduction project with just an `AudioStreamPlayer` playing a WAV audio. [Web Preview](https://davcri.github.io/godot-html5-thread-audio/) | [Repo ](https://github.com/davcri/godot-html5-thread-audio)

Also I can reproduce the issue simply by playing the WAV audio file with the official Godot Web Editor (see [video](https://youtu.be/LQYa5yoeI80)).

Tested with Godot 3.3 RC6 + Manjaro Linux + Chrome 89.0.4389.90. No issue with Firefox 87.0.

<details>
<summary>More HW/SW info: </summary>
<pre>
System:
  Kernel: 5.11.6-1-MANJARO x86_64 bits: 64 Desktop: Cinnamon 4.8.6 
  Distro: Manjaro Linux 
Machine:
  Type: Desktop System
  Mobo: ASUSTeK model: ROG STRIX B550-F GAMING (WI-FI) v: Rev X.0x 
CPU:
  Info: 8-Core model: AMD Ryzen 7 3700X bits: 64 type
Graphics:
  Device-1: NVIDIA TU116 [GeForce GTX 1650 SUPER] driver: nvidia v: 460.56 
  Display: x11 server: X.Org 1.20.10 driver: loaded: nvidia resolution: 
  1: 1920x1080~60Hz 2: 1600x900~60Hz 
  OpenGL: renderer: GeForce GTX 1650 SUPER/PCIe/SSE2 v: 4.6.0 NVIDIA 460.56 
Audio:
  Device-1: NVIDIA TU116 High Definition Audio driver: snd_hda_intel 
  Device-2: AMD Starship/Matisse HD Audio driver: snd_hda_intel 
  Device-3: RODE Microphones RODE NT-USB type: USB 
  driver: hid-generic,snd-usb-audio,usbhid 
  Sound Server-1: ALSA v: k5.11.6-1-MANJARO running: yes 
  Sound Server-2: PulseAudio v: 14.2 running: yes 
</pre>
</details>

**Comment by CedLec** [comment]
Hi,

Same issue for me when i export one my game as a HTML5 build. ( Video example : https://youtu.be/PMM2GxjyJqw )

**Comment by Calinou** [comment]
@davcri Can you reproduce this with an HTML5 export that has threads enabled in the export preset? Make sure to host the resulting project on a server that has [cross-origin isolation](https://web.dev/coop-coep/) headers enabled so that SharedArrayBuffer is available. (The project must also be accessed via either localhost or an HTTPS host.)

**Comment by davcri** [comment]
@Calinou thanks for the quick answer. The reproduction project that I posted before it's already using the thread-enabled HTML5 export.

Some findings:

- Issue it's still there with older Chrome version (80, 86)
- [Video](https://user-images.githubusercontent.com/6860637/112728048-9a497780-8f25-11eb-82f3-89cca3154ed1.mp4) | Issue may be related to a specific combination of Chrome + motherboard audio card + audio driver: audio it's pretty bad on the motherboard audio (`Starship/Matisse HD controller` which it's my daily output audio device), but it's almost okay with the headphones output (`RODE-NT USB` mic). However it still has some audio crackling. 
- [Video](https://user-images.githubusercontent.com/6860637/112728544-2066bd80-8f28-11eb-98e7-feeda116b8a4.mp4) | Single-thread export plays audio better than multi-thread export (which it's totally unexpected)
- :fox_face: Just to clarify: audio on Firefox it's almost perfect (with every combination of export and audio device)

**Comment by akien-mga** [comment]
Might be an issue with sampling rate, sounds related to #46595.

**Comment by Faless** [comment]
> Might be an issue with sampling rate, sounds related to #46595.

That's actually quite interesting, I've been suspecting issues with mismatching sample rates between godot and device for a while. It seems that most of the chrome devices on HTML5 that had problem was 48000 Hz (though changing it in Godot project settings to match it doesn't seem to consistently solve it), and on FF, I've been experiencing broken Mic support when sample rate does not match the default one (even if the chosen one is advertised as supported).

I wonder if we should default to "let device choose" since we seem to do resampling in real time anyway, and not during the import phase.

**Comment by 3p0ch-NG** [comment]
It affected me too. I had a Godot HTML build that was playing music fine until early March 2021, then suddenly started stuttering and getting delayed badly in Chrome and Edge, but not Firefox.

I don't know what's causing it, but I did find a workaround and posted a description of how I routed audio through howler.js in the spirit of FernandoValcazara's approach which completely eliminated the audio stuttering issue. The description of how I did it is at 
https://3p0ch.newgrounds.com/news/post/1148893

**Comment by Faless** [comment]
> It affected me too. I had a Godot HTML build that was playing music fine until early March 2021, then suddenly started stuttering and getting delayed badly in Chrome and Edge, but not Firefox.

So it seems a browser regression... can try which older version is the first to exhibit the bad behaviour?
That way we can report it to the chrome bug tracker: https://bugs.chromium.org/p/chromium/issues/list?q=component:Blink%3EWebAudio

**Comment by akien-mga** [comment]
Could also be an OS/drivers update that causes it, since @davcri could reproduce the regression on older Chrome too: https://github.com/godotengine/godot/issues/43210#issuecomment-808764511

**Comment by Faless** [comment]
Yeah, I'm quite convinced we are now dealing with 2 separate issues, a chrome regression, and the good old issues on some specific hardware (e.g.: [Use of AudioWorklets has glitches on some Windows systems](https://bugs.chromium.org/p/chromium/issues/detail?id=1188901&q=component%3ABlink%3EWebAudio) )

**Comment by Faless** [comment]
So, I can't open an upstream issue until someone that was not experiencing the issue before but is experiencing it now makes a proper report with **OS, hardware specs, first bogus version, last bogus version**.

Any generic "Was working, now since some days it doesn't", is just noise at this point.

**Comment by 3p0ch-NG** [comment]
> So, I can't open an upstream issue until someone that was not experiencing the issue before but is experiencing it now makes a proper report with **OS, hardware specs, first bogus version, last bogus version**.
> 
> Any generic "Was working, now since some days it doesn't", is just noise at this point.

I don't know how to rollback chromium and/or other updates to pinpoint the culprit. But if anyone does know how to, I posted a build that's completely empty except for a single player node playing a loop which reproduces the problem for me and could allow people to pinpoint it if they are also getting choppiness at https://www.newgrounds.com/portal/view/project/1594770 which might also help if more people can test it out on different systems. I have

Windows 10 Home, version 20H2, OS build 19042.867
Processor: AMD Ryzen 5 3500U with Radeon Vega Mobile Gfx, 2.10 GHz
Installed RAM: 12.0 GB (9.92 GB usable)
First and last bogus version: TBD if there's a straightforward way to determine it

**Comment by akien-mga** [comment]
> if they are also getting choppiness at https://www.newgrounds.com/portal/view/project/1594770 which might also help if more people can test it out on different systems.

Tested on Linux (Mageia 9) and I can reproduce choppy audio with Chrome 89.0.4389.90.

I tried to install older versions of Chromium using https://vikyd.github.io/download-chromium-history-version/#/ but at least the version I tried has its audio code crashing so it doesn't help :)

**Comment by akien-mga** [comment]
So I've done some testing on Windows 10 with Chrome 89.0.4389.90 on another laptop, same issue. I tried changing the audio driver sample rate for the Realtek card, but that doesn't seem to impact it (both 44100/48000 Hz and 16/24 Bits).

On Windows I managed to use the Chromium builds from https://vikyd.github.io/download-chromium-history-version/#/ but I can't seem to reproduce the issue with Chromium, only with Chrome :(

I tested the following builds:

- Chromium 88.0.4324.215 (latest build in previous stable): **works**
- Chromium 89.0.4326.0 (first build in 89.0 branch): **works**
- Chromium 89.0.4380.6: **works**
- Chromium 89.0.4386.4 (latest build in 89.0 branch at https://vikyd.github.io/download-chromium-history-version/#/): **works**
- Chromium 90.0.4391.0 (first build in 90.0 branch): **works**
- Chromium 91.0.4462.1 (dev version): **works**
- Chrome 89.0.4389.90 (current stable): **broken**

Fun if the regression is proprietary-only :@

**Comment by akien-mga** [comment]
@Faless Relevant bug reports?

https://bugs.chromium.org/p/chromium/issues/detail?id=1188901
https://bugs.chromium.org/p/chromium/issues/detail?id=1187016

**Comment by akien-mga** [comment]
> 
> 
> Yeah, I'm quite convinced we are now dealing with 2 separate issues, a chrome regression, and the good old issues on some specific hardware (e.g.: [Use of AudioWorklets has glitches on some Windows systems](https://bugs.chromium.org/p/chromium/issues/detail?id=1188901&q=component%3ABlink%3EWebAudio) )

Fabio is right, @davcri's issue with threaded builds is different from @3p0ch-NG's report about a Chrome 89 regression.

I've opened #47453 to keep track of the Chrome 89 regression with `ScriptProcessorNode`.

@davcri I suggest opening a new issue for your audio stuttering problem, which might be a manifestation of https://bugs.chromium.org/p/chromium/issues/detail?id=1188901 as Fabio pointed it out (AudioWorklet and not ScriptProcessorNode, and hardware-dependent). For the reference, I couldn't reproduce the issue with your hosted build on Chrome Stable or Canary on either Linux or Windows.

And by now this issue has been used to discussed quite a few unrelated problems so it's probably best if we close it, and focus on #47453 and a new issue for users hit by https://bugs.chromium.org/p/chromium/issues/detail?id=1188901.
